### DISCUSSION TITLE
MultiPhase modeling error...

### DISCUSSION THREAD

**Discussion by MScKarl** [idaholab/moose#23780] [Q&A Modules: Phase field] [answered]
Hi!

I am trying to use MOOSE to write a simulation for a two phase system, consisting of eta1 and eta2. But when they are defined as 

[Variables]
  [./eta1]
    order = FIRST
    family = LAGRANGE
  [../]

  [./eta2]
    order = FIRST
    family = LAGRANGE
  [../]

[]

[ICs]
  [./eta1]
    variable = eta1
    type = RandomIC
    min = 1e-5
    max = 1e-3
  [../]

  [./eta2]
    variable = eta2
    type = RandomIC
    min = 1e-5
    max = 1e-3
  [../]
[]

the following error occur:
 "*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Unknown variable 'eta1 eta2'. It does not exist in the nonlinear system(s) or auxiliary system ".

When I then tried using AuxVariables as following:
[AuxVariables]

  [./eta1]
    order = FIRST
    family = LAGRANGE
    [./InitialCondition]
      variable = eta1
      type = RandomIC
      min = 1e-5
      max = 1e-3
    [../]
  [../]

  [./eta2]
    order = FIRST
    family = LAGRANGE
    [./InitialCondition]
      variable = eta1
      type = RandomIC
      min = 1e-5
      max = 1e-3
    [../]
  [../]
[]

then I get the following error:
"*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

No nonlinear variable named eta1 found. Did you specify an auxiliary variable when you meant to specify a nonlinear variable?"

I would be very grateful if someone could tell me what I'm doing wrong and advise me how I should define my variables instead.

Best Regards
Karl

**Comment by GiudGiud** [comment, marked as answer]
Hello

these are two different errors from two different objects. 

For the first one, you specified two variables somewhere where only one was expected, so it’s confused at the name (thinks “eta1 eta2” is a variable name)

the variable definitions seem fine. If you want to use them as the variables to solve for they should be nonlinear

Guillaume

**Reply by MScKarl** [reply]
Thank you for your answer. Then I think I have found the problem, but it leads me to a question about understanding ACMultiPhase and its boundary conditions. If I use two eta variables to describe the different phases, how do I know which one to place where in the boundary conditions? Previously I wrote this:

[BCs]

  [./left]
    type = NeumannBC
    variable = 'eta1 eta2'
    boundary = 'left'
    value = 0.0
  [../]

  [./right]
    type = NeumannBC
    variable = 'eta1 eta2'
    boundary = 'right'
    value = 0.0
  [../]

  [./top]
    type = NeumannBC
    variable = 'eta1 eta2'
    boundary = 'top'
    value = 0.0
  [../]

  [./bottom]
    type = NeumannBC
    variable = 'eta1 eta2'
    boundary = 'bottom'
    value = 0.0
  [../]

[]

but if  I understood your answer it should only have one. Therfore I wonder if, depending on where in the BC I place my eta1 or eta2 will that determine where the different phases will be placed in the system? Or how do use the NeumannBC when there are two etas?

/Karl

**Reply by GiudGiud** [reply]
Hello

You can duplicate safely the boundary conditions. For example:
```
[./bottom_1]
type = NeumannBC
variable = 'eta1'
boundary = 'bottom'
value = 0.0
[../]
[./bottom_2]
type = NeumannBC
variable = 'eta2'
boundary = 'bottom'
value = 0.0
[../]
```
are you worried a phase is not present near a boundary?
If that happens, the boundary condition will just not trigger

**Reply by MScKarl** [reply]
Great! Thanks.


if it's not too much of an inconvenience may I ask another question?

by using the aforementioned definition of eta1 & eta 2, and the material below, why doesn't moose accept my function in [./fa] and [./fb]?

```
[Materials]
  [./fa]
    type = DerivativeParsedMaterial
    property_name = fa
    coupled_variables = 'eta1'
    function = (eta1^2-1)^2
  [../]

  [./fb]
    type = DerivativeParsedMaterial
    f_name = fb
    args = 'eta2'
    function = '(eta2^2-1)^2'
  [../]
[]

**Reply by GiudGiud** [reply]
You have a mix of the new and old syntax for parameters here. 
what errors do you get?

**Reply by MScKarl** [reply]
I get the following error:

*** ERROR ***
Invalid function
(eta1^2-1)^2

in ParsedMaterialHelper.
Syntax error: Unknown identifier

**Reply by GiudGiud** [reply]
oh
You need to update MOOSE to use the new parameters

right now it doesnt know what you mean by eta1 because it expects args=eta1 not coupled_variables = eta1

**Reply by MScKarl** [reply]
I've tried to upgrade my MOOSE system, and  then tried both 

(eta1^2-1)^2

and

(eta1 * eta1 - 1) * (eta1 * eta1 - 1)

which both results in the same error.

**Reply by GiudGiud** [reply]
did you recompile?
what does `git log` say in the moose folder?

**Reply by MScKarl** [reply]
Restarted the sytem, and now the error is gone. I apologies for the inconvenience.

Again, thank for all your help. It is very much appreachiated.
